flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -503,12 +508,11 @@ private void onGlobalPoolAvailable() {
         mayNotifyAvailable(toNotify);
     }
 
+    @GuardedBy("availableMemorySegments")
     private boolean shouldBeAvailable() {
         assert Thread.holdsLock(availableMemorySegments);
 
-        return !availableMemorySegments.isEmpty()
-                && unavailableSubpartitionsCount == 0
-                && numberOfRequestedOverdraftMemorySegments == 0;
+        return !availableMemorySegments.isEmpty() && unavailableSubpartitionsCount == 0;
```

Comment:
Why the `numberOfRequestedOverdraftMemorySegments  == 0` can be removed? I think there is a constraint here: The `availableMemorySegments` is always empty when `numberOfRequestedOverdraftMemorySegments > 0`. 

Sometimes this constraint does not hold before this PR, and we want to hold the constraint in the future, right?

If yes, could you add some comments for `numberOfRequestedOverdraftMemorySegments`? It's helpful for other developers to understand the constraint.

Reply:
Before we allows convert  `numberOfRequestedOverdraftMemorySegments` to `numberOfRequestedMemorySegments `, there still be a situation where both `availableMemorySegment` and `overdraft buffer` are all not zero, and this state should obviously be defined as available. But after that, I'm not sure whether this situation still exists. I'm a little worried about the potential bug of multithreading if we 
 add `numberOfRequestedOverdraftMemorySegments  == 0` back to `shouldBeAvailable`. 
For me, the key point is that if we think "`The availableMemorySegments` is always empty when `numberOfRequestedOverdraftMemorySegments` != 0." is tenable, does `!availableMemorySegments.isEmpty()` already include `numberOfRequestedOverdraftMemorySegments == 0`? Even if this will not introduce bug, why should we impose useless constraints? 
Actually, before overdraft buffer was introduced, the definition of `available` was very clear:
There is at least one `availableMemorySegment` and no subpartitions has reached `maxBuffersPerChannel`.
IMO, Introducing the overdraft mechanism should not break this protocol, overdraft buffer should not affect the judgment of availability.

Reply:
> Actually, before overdraft buffer was introduced, the definition of available was very clear:
There is at least one availableMemorySegment and no subpartitions has reached maxBuffersPerChannel.
IMO, Introducing the overdraft mechanism should not break this protocol, overdraft buffer should not affect the judgment of availability.

Sorry, overdraft buffer should affect the judgment of availability.

As we discussed before: overdraft buffer is just used when requested buffer reached the upper limit(pool size). In the other word: overdraft buffer should be used after LocalBufferPool is unavailable. 

And why the name is `overdraft`? It temporarily uses some extra buffers outside the LocalBufferPool. From the semantics of overdraft, if `numberOfRequestedOverdraftMemorySegments > 0`, then LocalBufferPool must be unavailable. That's why I add it here.

Why you want to remove it? I guess it has bug before, that is, the overdraft buffer is used when the `requested buffer` does not reach the upper limit, and you have fixed it in this PR.

```
            if (!availableMemorySegments.isEmpty()) {
                segment = availableMemorySegments.poll();
            } else if (isRequestedSizeReached()) {
                // Only when the buffer request reaches the upper limit(i.e. current pool size),
                // requests an overdraft buffer.
                segment = requestOverdraftMemorySegmentFromGlobal();
            }
```

> For me, the key point is that if we think "The availableMemorySegments is always empty when numberOfRequestedOverdraftMemorySegments != 0." is tenable.

If it's tenable now and future, we can remove the ` && numberOfRequestedOverdraftMemorySegments == 0` here, if not, it cannot be removed.

Based on your feedback, I prefer keep it, because the root cause is: overdraft is misused in some cases, and you have fixed it.

If there is other bug that misuse overdraft buffer, we should fix the bug to ensure overdraft buffer is used correctly instead of mark LocalBufferPool is available. Marking the LocalBufferPool is available directly may cause other unexpected bugs.

Reply:
> Before we allows convert numberOfRequestedOverdraftMemorySegments to numberOfRequestedMemorySegments , there still be a situation where both availableMemorySegment and overdraft buffer are all not zero.

Just as what I said, only after we supports convert `numberOfRequestedOverdraftMemorySegments` to `numberOfRequestedMemorySegments`, then can we avoid having both `available buffers` and `overdraft buffers`. 
It is not enough to just modify the following code as there are still concurrency problems.
```
            if (!availableMemorySegments.isEmpty()) {
                segment = availableMemorySegments.poll();
            } else if (isRequestedSizeReached()) {
                // Only when the buffer request reaches the upper limit(i.e. current pool size),
                // requests an overdraft buffer.
                segment = requestOverdraftMemorySegmentFromGlobal();
            }
```
Of course, we have allowed this conversion now, but I'm not very sure whether there are other concurrency problems even though after this fix.
In any case, your reason is also make sense and acceptable to me. But for the sake of safety, perhaps we should listen to the opinions of @wsry who knows more about the implementation of `LocalBufferPool` than me.

Reply:
Hi @pnowojski @akalash , could you help take a look in your free time, thanks

Reply:
I actually mostly agree that `if (isRequestedSizeReached())` condition and `while` inside `setNumBuffers` fix the bug of this ticket. It is difficult to say something about removing `numberOfRequestedOverdraftMemorySegments == 0`. On one hand, it is better doesn't change anything and improve it in another PR(if we still think it makes sense) since it doesn't relate to this bug, on the other hand, in the worst case(if we still have a bug) removing this condition just allows us to allocate 'ordinary' buffer while we have already overdraft buffers which is not so good but still better than have a deadlock. Let me think about it a little more. But anyway if we will decide to remove it I think it makes sense to do it in a separate commit just to understand where is the fix and where is the prophylactic/improvement

In general, I would say that it seems we need to improve/refactor this class a little(maybe I will create a ticket for that). Since I think we don't have a strong/explicit relationship between  `numberOfRequestedOverdraftMemorySegments`, `numberOfRequestedMemorySegments`, `availableMemorySegment` and `currentPoolSize` for example:
- `!availableMemorySegment.isEmpty() && numberOfRequestedOverdraftMemorySegments != 0` - never should be true
- `numberOfRequestedMemorySegments == currentPoolSize && availableMemorySegment.isEmpty()` - always should be true when `numberOfRequestedOverdraftMemorySegments > 0`
- etc.

Reply:
Thanks @akalash for the timely feedback.

It makes sense to remove `numberOfRequestedOverdraftMemorySegments == 0` in another PR. Have to say, I like your suggestion about refactoring this class as the current logic looks bloated and a little confusing. Once we can sort out the relationship between these concepts, it can also help us understand whether we need to keep `numberOfRequestedOverdraftMemorySegments == 0`. 

I will keep this condition in this PR first and take a closer look at whether it's necessary to remove it when the refactoring is complete.

Reply:
Hi @akalash @reswqa , thanks for your quick feedback. Sounds make sense.

@reswqa could you squash the last commit to the [FLINK-31293] commit?

Reply:
@1996fanrui Sure, the reason why I use fix-up is to make it easier for everyone to see the changes I made last time. If there are no other issues, I'll squash it.